### PR TITLE
{List,Create,Change}InputLogs Admin APIs

### LINKED ISSUE #1063: Create tool for adding / removing queue shards at runtime
The key server supports writing to multiple shards for horizontal scaling of incoming requests. 

This is currently set to a static number during initial domain creation.

We should create an API so an administrator can turn shards up/down as they please.

### PR DESCRIPTION
**Pull request by gdbelvin**
Support admin control of input log horizontal scaling.

Fixes #1063

### CODE REVIEW

File: core/adminserver/admin_server.go [hunk 1]
```diff
@@ -83,8 +83,14 @@ var (
 type LogsAdmin interface {
 	// AddLogs creates and adds new logs for writing to a directory.
 	AddLogs(ctx context.Context, directoryID string, logIDs ...int64) error
+	// SetWritable enables or disables new writes from going to logID.
+	SetWritable(ctx context.Context, directoryID string, logID int64, enabled bool) error
+	// ListLogs returns a list of logs, optionally filtered by the writable bit.
+	ListLogs(ctx context.Context, directoryID string, writable bool) ([]int64, error)
 }
 
+var _ pb.KeyTransparencyAdminServer = &Server{}
```
Comment (therealdrake) [inline core/adminserver/admin_server.go]: what is this line doing?

File: core/adminserver/admin_server.go [hunk 2]
```diff
@@ -213,11 +219,13 @@ func (s *Server) CreateDirectory(ctx context.Context, in *pb.CreateDirectoryRequ
 	// Generate VRF key.
 	wrapped, err := privKeyOrGen(ctx, in.GetVrfPrivateKey(), s.keygen)
 	if err != nil {
```
Comment (therealdrake) [inline core/adminserver/admin_server.go]: could this just be:
if code := status.Code(err); code != codes.OK {
  return nil, status.Errorf(code, "adminserver: keygen(): %v", err)
}

File: core/adminserver/admin_server.go [hunk 3]
```diff
@@ -370,6 +389,39 @@ func (s *Server) UndeleteDirectory(ctx context.Context, in *pb.UndeleteDirectory
...
 }
 
+// ListInputLogs returns a list of input logs for a directory.
+func (s *Server) ListInputLogs(ctx context.Context, in *pb.ListInputLogsRequest) (*pb.ListInputLogsResponse, error) {
+	logIDs, err := s.logsAdmin.ListLogs(ctx, in.GetDirectoryId(), in.GetFilterWritable())
+	if err != nil {
+		s := status.Convert(err)
+		return nil, status.Errorf(s.Code(), "adminserver: ListLogs(): %v", s.Message())
+	}
+	inputLogs := make([]*pb.InputLog, 0, len(logIDs))
```
Comment (therealdrake) [inline core/adminserver/admin_server.go]: I don't know how much better this is than:
inputLogs := []*pb.InputLog{}
which I find more readable

File: core/adminserver/admin_server.go [hunk 4]
```diff
@@ -370,6 +389,39 @@ func (s *Server) UndeleteDirectory(ctx context.Context, in *pb.UndeleteDirectory
...
+	for _, logID := range logIDs {
+		inputLogs = append(inputLogs, &pb.InputLog{LogId: logID, Writable: true})
+	}
+
+	return &pb.ListInputLogsResponse{Logs: inputLogs}, nil
+}
+
+// CreateInputLog returns a the created log.
```
Comment (therealdrake) [inline core/adminserver/admin_server.go]: "a" or "the"

File: impl/sql/mutationstorage/queue_test.go [hunk 5]
```diff
@@ -40,6 +40,52 @@ func newForTest(ctx context.Context, t testing.TB, logIDs ...int64) *Mutations {
...
+			}
+
+			for logID, enabled := range tc.set {
+				if err := m.SetWritable(ctx, directoryID, logID, enabled); err != nil {
+					t.Errorf("SetWritable(): %v", err)
+				}
+			}
+
+			logs := make(map[int64]bool)
+			// Run enough times to ensure that random sampling will get us all logs.
```
Comment (therealdrake) [inline impl/sql/mutationstorage/queue_test.go]: could you instead set all logs not-writable except for one and ensure that is the one that gets written to?

File: core/adminserver/admin_server.go (hunk 1 above)
Comment (gdbelvin) [reply core/adminserver/admin_server.go]: This line ensures that `&Server{}` implements the `pb.KeyTransparencyAdminServer` interface.

Comment (gdbelvin) [reply core/adminserver/admin_server.go]: I'll add a comment to that effect.

File: core/adminserver/admin_server.go (hunk 2 above)
Comment (gdbelvin) [reply core/adminserver/admin_server.go]: hmm, that would be better...

File: core/adminserver/admin_server.go (hunk 3 above)
Comment (gdbelvin) [reply core/adminserver/admin_server.go]: The only difference is that the way it's written pre-allocates memory so we can avoid multiple runtime array resizing.

File: impl/sql/mutationstorage/queue_test.go (hunk 5 above)
Comment (gdbelvin) [reply impl/sql/mutationstorage/queue_test.go]: Sure

File: core/adminserver/admin_server.go (hunk 1 above)
Comment (therealdrake) [reply core/adminserver/admin_server.go]: ah that's clever ok

**Comment by gdbelvin** [comment]
Updated. PTAL

File: core/adminserver/admin_server.go [hunk 6]
```diff
@@ -404,19 +390,17 @@ func (s *Server) ListInputLogs(ctx context.Context, in *pb.ListInputLogsRequest)
...
 		return nil, status.Errorf(s.Code(), "adminserver: AddLogs(%+v): %v", in.GetLogId(), s.Message())
 	}
 	return &pb.InputLog{LogId: in.GetLogId(), Writable: true}, nil
 }
 
 // UpdateInputLog updates the write bit for an input log.
 func (s *Server) UpdateInputLog(ctx context.Context, in *pb.InputLog) (*pb.InputLog, error) {
-	if err := s.logsAdmin.SetWritable(ctx, in.GetDirectoryId(), in.GetLogId(), in.GetWritable()); err != nil {
-		s := status.Convert(err)
+	if s := status.Convert(s.logsAdmin.SetWritable(ctx, in.GetDirectoryId(), in.GetLogId(), in.GetWritable())); s.Code() != codes.OK {
```
Comment (therealdrake) [inline core/adminserver/admin_server.go]: I feel like the call to SetWritable should go on the line above

File: core/adminserver/admin_server.go [hunk 7]
```diff
@@ -404,19 +390,17 @@ func (s *Server) ListInputLogs(ctx context.Context, in *pb.ListInputLogsRequest)
 	return &pb.ListInputLogsResponse{Logs: inputLogs}, nil
 }
 
-// CreateInputLog returns a the created log.
+// CreateInputLog returns the created log.
 func (s *Server) CreateInputLog(ctx context.Context, in *pb.InputLog) (*pb.InputLog, error) {
-	if err := s.logsAdmin.AddLogs(ctx, in.GetDirectoryId(), in.GetLogId()); err != nil {
-		s := status.Convert(err)
+	if s := status.Convert(s.logsAdmin.AddLogs(ctx, in.GetDirectoryId(), in.GetLogId())); s.Code() != codes.OK {
```
Comment (therealdrake) [inline core/adminserver/admin_server.go]: I feel like the call to AddLogs should go on the line above

**Pull request merged**